### ISSUE TITLE
Translations on Loco are overwritten with translation keys

### ISSUE THREAD

**Issue by nicwortel** [php-translation/symfony-bundle#280]
I'm still trying to figure out what is going on, so I'm not sure if this issue resides in this bundle or in the Loco adapter. I wasn't able to find an existing issue.

We're using XLIFF (2.0 if that matters) in a Symfony 4 application. When running `bin/console translation:sync default down`, the existing translations in Loco get overwritten with translation keys:

![afbeelding](https://user-images.githubusercontent.com/1055691/52576156-ddb4fa80-2e1f-11e9-948a-aff9138b7744.png)

It seems to be a very specific issue because only the dutch translations are overwritten, other translations seem to be unaffected. Locally, the `srcLang` attribute of the XLIFF files is changed from "en" to "nl-NL":

![afbeelding](https://user-images.githubusercontent.com/1055691/52576451-60d65080-2e20-11e9-8d8d-e1d1f8b6a103.png)

My computer's locale is "nl-NL", so perhaps that has something to do with it. I see that `XliffDumper` from the Symfony Storage package defaults to `Locale::getDefault()` for the `$defaultLocale` value. However, I haven't found out yet why that overwrites the translations on Loco.

We're using the following versions, which as far as I'm aware are the latest:

```
$ composer show php-translation/*
php-translation/common          1.0.0 Common translation stuff
php-translation/extractor       1.7.1 Extract translations form the source code
php-translation/loco-adapter    0.8.0 Adapter for loco.
php-translation/symfony-bundle  0.8.1
php-translation/symfony-storage 1.0.1 A translation file storage using Symfony translation component.
```

**Comment by nicwortel** [comment]
After investigating this issue a bit more, I believe that I have found the source of the problem. It appears to be twofold, caused by a combination of issues in Loco and in the Loco adapter.

The `import()` method of the `Loco` class (used by the [`translation:sync` command](https://github.com/php-translation/symfony-bundle/blob/master/Command/SyncCommand.php)) [calls `XliffConverter::catalogueToContent()` without the `$options` array](https://github.com/php-translation/loco-adapter/blob/ef9c3a0a966ca8ca3f3ecdb69264bb974fbdadc7/src/Loco.php#L198). The `catalogueToContent()` method then uses the [`XliffDumper`](https://github.com/php-translation/symfony-storage/blob/master/src/Dumper/XliffDumper.php) to format the catalogue as an XLIFF file, [passing along the `$options` array](https://github.com/php-translation/symfony-storage/blob/8d7b1ea6bf22bee0c7dc9b228cf9437312373a27/src/XliffConverter.php#L59) with an additional `xliff_version` key set to `2.0`.

The `XliffDumper` then tries to determine the "source locale" to be used in the dumped XLIFF file. [If the `$options` array contains a key `default_locale` it uses that, and in all other cases it falls back to `\Locale::getDefault()`](https://github.com/php-translation/symfony-storage/blob/8d7b1ea6bf22bee0c7dc9b228cf9437312373a27/src/Dumper/XliffDumper.php#L45-L49) (this logic seems to be copied from [Symfony's `XliffFileDumper`](https://github.com/symfony/symfony/blob/6a80e7ddd5e476a2bda88913d6ae041edf389248/src/Symfony/Component/Translation/Dumper/XliffFileDumper.php#L34-L38)). So because at this point `$options` only contains `xliff_version`, `$defaultLocale` is set to my system locale, which is "nl_NL".

After this, the `$defaultLocale` is passed to the dumper that is actually used to create the XLIFF file - which can either be Symfony's XliffFileDumper (for XLIFF 1.2) or [`SymfonyPort`](https://github.com/php-translation/symfony-storage/blob/master/src/Dumper/Port/SymfonyPort.php) (for XLIFF 2.0), to be used in the XLIFF file as the `source-language` (1.2) or `srcLang` (2.0) attribute. In my case this explains why the `srcLang` in my XLIFF files is being changed from "en" to "nl-NL".

The second part of the issue seems to be with Loco. I have been able to verify this by manually importing XLIFF files through Loco's user interface. If I import an XLIFF file with `srcLang="en"`, everything works fine. However, if I then change the `srcLang` attribute to "nl-NL", and import the file again, all my (Dutch) translations are overwritten with the translation keys.

![afbeelding](https://user-images.githubusercontent.com/1055691/52788750-31af2180-3062-11e9-8571-d11c9a9debcb.png)

From my perspective, it seems that both the Loco adapter / this bundle and Loco itself are wrong:

1. The Loco adapter should pass the `default_locale` option to the XliffConverter instead of letting it guess it based on the developer's locale. If `index_parameter` is set to `id`, it probably makes sense to set `default_locale` to "en"?
2. Loco should respect the fact that I chose "Keys are asset IDs", instead of using the `<source>` value instead of `<target>` which effectively overwrites my translations with their translation keys.

I will report the second issue to Loco, because it seems like that is the main issue here. However, I feel like the first issue should be addressed as well, in the Loco adapter. @Nyholm how do you feel abou this? Do you want me to create an issue in https://github.com/php-translation/loco-adapter?

**Comment by timwhitlock** [comment]
Regarding my part in all of this:

> Loco should respect the fact that I chose "Keys are asset IDs", instead of using the `<source>` value instead of `<target>`

Your target translations will normally be pulled from the `<target>` element, unless the `srcLang` attribute is a **better** match for the language you've specified.  

Example: `srcLang="nl-NL" trgLang="nl"`

Here `nl-NL` is more specific and matches exactly the locale in your Loco project, so the importer thinks you are asking it to select the `<source>` element on purpose.

It's reasonable to expect the "Keys are asset IDs" option would effectively switch off this selection behaviour, but XLIFF (being more complex than simple key/value pairs) means that the two processes are separate and can even be contradictory. Basically the Loco importer is saying this:

1. I see your `"name" `attribute. That is your key and I will NOT use is as a source text.
2. I see your `<source>` element. It matches the language you want better than the `<target>` does, so I'll use that as your translation.

**Comment by timwhitlock** [comment]
I wonder if @Nyholm would consider using `srcLang="zxx"` (or similar) when the `<source>` element contains language-agnostic keys.

**Comment by nicwortel** [comment]
@timwhitlock thanks again for your support. Is this behaviour documented in Loco's documentation?

Given that this behaviour of Loco is intended, I think the Loco adapter should dump the XLIFF file with a fixed `srcLang` - either "en" or a non-existing language code (something like "zxx") should work if I'm not mistaken, although it sounds like the latter is more correct.

I'll create a pull request in the [php-translation/loco-adapter](https://github.com/php-translation/loco-adapter) repository.

**Comment by timwhitlock** [comment]
> Is this behaviour documented in Loco's documentation?

There are some [Symfony notes here](https://localise.biz/help/formats/importing/xliff#symfony) but I will add a warning about source/target selection. 

> I think the Loco adapter should dump the XLIFF file with a fixed `srcLang`

I agree, but I've been thinking about a simple change I could make to the API too. 

Instead of specifying `?locale=nl-NL`, I'm considering an option like `locale=target` or `locale=auto`. This could automatically select the file's *target* language without the risk of ambiguity as shown here. I'll look at this today.

**Comment by timwhitlock** [comment]
Sorry for the delay. I've just pushed an update to the API that allows you to specify `locale=target`, which will select only the `<target>` element and import it according to the `trgLang` attribute.

This is not yet available in the user interface, but will be shortly.

The [Symfony notes](https://localise.biz/help/formats/symfony) have been updated with this as the recommended option, and the parameter itself is documented in more detail in [Importing XLIFF files](https://localise.biz/help/formats/importing/xliff).
